File: mytravelstatus/mobile/android/app/src/main/java/com/mytravelstatus/app/service/SubscriptionManager.kt
/**
 * SubscriptionManager.kt
 *
 * Google Play Billing subscription management.
 * Implements auto-renewable subscriptions with:
 * - Full pricing display with period
 * - Auto-renewal terms disclosure
 * - Restore purchases
 * - Manage subscription link
 *
 * @package MyTravelStatus
 * @since   1.0.0
 */

package com.mytravelstatus.app.service

import android.app.Activity
import android.content.Context
import android.util.Log
import com.android.billingclient.api.*
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * Subscription product IDs
 */
object SubscriptionProducts {
    const val PREMIUM_MONTHLY = "com.mytravelstatus.premium.monthly"
    const val PREMIUM_ANNUAL = "com.mytravelstatus.premium.annual"

    val ALL_PRODUCTS = listOf(PREMIUM_MONTHLY, PREMIUM_ANNUAL)
}

/**
 * Subscription status
 */
data class SubscriptionStatus(
    val isActive: Boolean = false,
    val productId: String? = null,
    val expiryDate: Long? = null,
    val isAutoRenewing: Boolean = false,
    val displayText: String = "Free Plan"
)

object SubscriptionManager : PurchasesUpdatedListener {
    private const val TAG = "SubscriptionManager"

    private var billingClient: BillingClient? = null
    private val coroutineScope = CoroutineScope(Dispatchers.Main + SupervisorJob())

    // StateFlows for reactive UI updates
    private val _products = MutableStateFlow<List<ProductDetails>>(emptyList())
    val products: StateFlow<List<ProductDetails>> = _products.asStateFlow()

    private val _subscriptionStatus = MutableStateFlow(SubscriptionStatus())
    val subscriptionStatus: StateFlow<SubscriptionStatus> = _subscriptionStatus.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _purchaseError = MutableStateFlow<String?>(null)
    val purchaseError: StateFlow<String?> = _purchaseError.asStateFlow()

    /**
     * Initialize the billing client
     */
    fun initialize(context: Context) {
        if (billingClient != null) return

        billingClient = BillingClient.newBuilder(context)
            .setListener(this)
            .enablePendingPurchases()
            .build()

        startConnection()
    }

    /**
     * Start billing client connection
     */
    private fun startConnection() {
        billingClient?.startConnection(object : BillingClientStateListener {
            override fun onBillingSetupFinished(result: BillingResult) {
                if (result.responseCode == BillingClient.BillingResponseCode.OK) {
                    Log.d(TAG, "Billing client connected")
                    coroutineScope.launch {
                        queryProducts()
                        queryPurchases()
                    }
                } else {
                    Log.e(TAG, "Billing setup failed: ${result.debugMessage}")
                }
            }

            override fun onBillingServiceDisconnected() {
                Log.d(TAG, "Billing client disconnected")
                // Try to reconnect
                coroutineScope.launch {
                    delay(5000)
                    startConnection()
                }
            }
        })
    }

    /**
     * Query available subscription products
     */
    private suspend fun queryProducts() {
        val client = billingClient ?: return

        val productList = SubscriptionProducts.ALL_PRODUCTS.map { productId ->
            QueryProductDetailsParams.Product.newBuilder()
                .setProductId(productId)
                .setProductType(BillingClient.ProductType.SUBS)
                .build()
        }

        val params = QueryProductDetailsParams.newBuilder()
            .setProductList(productList)
            .build()

        val result = client.queryProductDetails(params)

        if (result.billingResult.responseCode == BillingClient.BillingResponseCode.OK) {
            _products.value = result.productDetailsList ?: emptyList()
            Log.d(TAG, "Products loaded: ${_products.value.size}")
        } else {
            Log.e(TAG, "Failed to query products: ${result.billingResult.debugMessage}")
        }
    }

    /**
     * Query existing purchases to check subscription status
     */
    private suspend fun queryPurchases() {
        val client = billingClient ?: return

        val params = QueryPurchasesParams.newBuilder()
            .setProductType(BillingClient.ProductType.SUBS)
            .build()

        val result = client.queryPurchasesAsync(params)

        if (result.billingResult.responseCode == BillingClient.BillingResponseCode.OK) {
            val activePurchase = result.purchasesList.find { purchase ->
                purchase.purchaseState == Purchase.PurchaseState.PURCHASED
            }

            if (activePurchase != null) {
                updateSubscriptionStatus(activePurchase)
            } else {
                _subscriptionStatus.value = SubscriptionStatus()
            }
        }
    }

    /**
     * Launch purchase flow for a product
     */
    fun purchase(activity: Activity, productDetails: ProductDetails) {
        val client = billingClient
        if (client == null) {
            _purchaseError.value = "Billing not initialized"
            return
        }

        _isLoading.value = true
        _purchaseError.value = null

        // Get the subscription offer
        val offerToken = productDetails.subscriptionOfferDetails?.firstOrNull()?.offerToken
        if (offerToken == null) {
            _purchaseError.value = "No subscription offer available"
            _isLoading.value = false
            return
        }

        val productDetailsParams = BillingFlowParams.ProductDetailsParams.newBuilder()
            .setProductDetails(productDetails)
            .setOfferToken(offerToken)
            .build()

        val params = BillingFlowParams.newBuilder()
            .setProductDetailsParamsList(listOf(productDetailsParams))
            .build()

        val result = client.launchBillingFlow(activity, params)

        if (result.responseCode != BillingClient.BillingResponseCode.OK) {
            _purchaseError.value = "Failed to launch purchase: ${result.debugMessage}"
            _isLoading.value = false
        }

        AnalyticsManager.track(AnalyticsEvent.SUBSCRIPTION_PURCHASE_STARTED, mapOf(
            "subscription_id" to productDetails.productId
        ))
    }

    /**
     * Restore purchases
     */
    fun restorePurchases() {
        coroutineScope.launch {
            _isLoading.value = true
            queryPurchases()
            _isLoading.value = false
            AnalyticsManager.track(AnalyticsEvent.SUBSCRIPTION_RESTORED)
        }
    }

    /**
     * Open subscription management in Play Store
     */
    fun openManageSubscription(context: Context) {
        val productId = _subscriptionStatus.value.productId ?: return

        val url = "https://play.google.com/store/account/subscriptions?sku=$productId&package=${context.packageName}"
        val intent = android.content.Intent(android.content.Intent.ACTION_VIEW)
        intent.data = android.net.Uri.parse(url)
        intent.flags = android.content.Intent.FLAG_ACTIVITY_NEW_TASK
        context.startActivity(intent)
    }

    /**
     * Get the auto-renewal disclosure text for a product (REQUIRED by Google Play)
     */
    fun getRenewalDisclosure(productDetails: ProductDetails): String {
        val offer = productDetails.subscriptionOfferDetails?.firstOrNull()
        val pricingPhase = offer?.pricingPhases?.pricingPhaseList?.firstOrNull()

        val price = pricingPhase?.formattedPrice ?: "N/A"
        val period = when (pricingPhase?.billingPeriod) {
            "P1M" -> "month"
            "P1Y" -> "year"
            "P1W" -> "week"
            else -> "period"
        }

        return """
            Subscription automatically renews at $price/$period unless canceled at least 24 hours before the end of the current period. Your Google Play account will be charged for renewal within 24 hours prior to the end of the current period. You can manage and cancel your subscription in your Google Play Store account settings after purchase.
        """.trimIndent()
    }

    /**
     * Get formatted price for display
     */
    fun getFormattedPrice(productDetails: ProductDetails): String {
        return productDetails.subscriptionOfferDetails?.firstOrNull()
            ?.pricingPhases?.pricingPhaseList?.firstOrNull()
            ?.formattedPrice ?: "N/A"
    }

    /**
     * Get billing period display text
     */
    fun getBillingPeriod(productDetails: ProductDetails): String {
        val period = productDetails.subscriptionOfferDetails?.firstOrNull()
            ?.pricingPhases?.pricingPhaseList?.firstOrNull()
            ?.billingPeriod

        return when (period) {
            "P1M" -> "month"
            "P1Y" -> "year"
            "P1W" -> "week"
            else -> "period"
        }
    }

    /**
     * Check if user has active premium subscription
     */
    fun isPremium(): Boolean {
        // In demo mode, always return true
        if (DemoModeService.isDemoMode.value) {
            return true
        }
        return _subscriptionStatus.value.isActive
    }

    // MARK: - PurchasesUpdatedListener

    override fun onPurchasesUpdated(result: BillingResult, purchases: List<Purchase>?) {
        _isLoading.value = false

        when (result.responseCode) {
            BillingClient.BillingResponseCode.OK -> {
                purchases?.forEach { purchase ->
                    handlePurchase(purchase)
                }
            }
            BillingClient.BillingResponseCode.USER_CANCELED -> {
                Log.d(TAG, "Purchase canceled by user")
                _purchaseError.value = null // User intentionally canceled
            }
            else -> {
                Log.e(TAG, "Purchase failed: ${result.debugMessage}")
                _purchaseError.value = "Purchase failed: ${result.debugMessage}"
                AnalyticsManager.track(AnalyticsEvent.SUBSCRIPTION_PURCHASE_FAILED)
            }
        }
    }

    // MARK: - Private Methods

    private fun handlePurchase(purchase: Purchase) {
        if (purchase.purchaseState == Purchase.PurchaseState.PURCHASED) {
            // Acknowledge the purchase if not already acknowledged
            if (!purchase.isAcknowledged) {
                coroutineScope.launch {
                    acknowledgePurchase(purchase)
                }
            }

            updateSubscriptionStatus(purchase)
            AnalyticsManager.track(AnalyticsEvent.SUBSCRIPTION_PURCHASE_COMPLETED, mapOf(
                "subscription_id" to (purchase.products.firstOrNull() ?: "unknown")
            ))
        }
    }

    private suspend fun acknowledgePurchase(purchase: Purchase) {
        val client = billingClient ?: return

        val params = AcknowledgePurchaseParams.newBuilder()
            .setPurchaseToken(purchase.purchaseToken)
            .build()

        val result = client.acknowledgePurchase(params)
        if (result.responseCode == BillingClient.BillingResponseCode.OK) {
            Log.d(TAG, "Purchase acknowledged")
        } else {
            Log.e(TAG, "Failed to acknowledge purchase: ${result.debugMessage}")
        }
    }

    private fun updateSubscriptionStatus(purchase: Purchase) {
        val productId = purchase.products.firstOrNull()
        val product = _products.value.find { it.productId == productId }

        val periodText = product?.let { getBillingPeriod(it) } ?: "subscription"

        _subscriptionStatus.value = SubscriptionStatus(
            isActive = true,
            productId = productId,
            expiryDate = null, // Would need to query from server
            isAutoRenewing = purchase.isAutoRenewing,
            displayText = "Premium ($periodText)"
        )
    }

    /**
     * Clean up resources
     */
    fun destroy() {
        billingClient?.endConnection()
        billingClient = null
        coroutineScope.cancel()
    }
}

File: mytravelstatus/mobile/ios/MyTravelStatus/Services/SubscriptionManager.swift
/**
 * SubscriptionManager.swift
 *
 * StoreKit 2 subscription management for MyTravelStatus.
 * Features:
 * - Auto-renewable subscriptions
 * - Restore purchases
 * - Manage subscription links
 * - Full pricing/renewal disclosures
 * - App Store review compliance
 *
 * @package MyTravelStatus
 * @since   1.0.0
 */

import Foundation
import StoreKit

/// Subscription tier
enum SubscriptionTier: String, CaseIterable {
    case monthly = "com.mytravelstatus.premium.monthly"
    case annual = "com.mytravelstatus.premium.annual"

    var displayName: String {
        switch self {
        case .monthly: return "Monthly"
        case .annual: return "Annual"
        }
    }

    var period: String {
        switch self {
        case .monthly: return "month"
        case .annual: return "year"
        }
    }
}

/// Subscription status
enum SubscriptionStatus: Equatable {
    case notSubscribed
    case subscribed(tier: SubscriptionTier, expirationDate: Date, willRenew: Bool)
    case expired(tier: SubscriptionTier, expirationDate: Date)
    case inGracePeriod(tier: SubscriptionTier, expirationDate: Date)
    case inBillingRetry(tier: SubscriptionTier)

    var isActive: Bool {
        switch self {
        case .subscribed, .inGracePeriod, .inBillingRetry:
            return true
        case .notSubscribed, .expired:
            return false
        }
    }

    var displayText: String {
        switch self {
        case .notSubscribed:
            return "Not subscribed"
        case .subscribed(let tier, let date, let willRenew):
            let formatter = DateFormatter()
            formatter.dateStyle = .medium
            let renewal = willRenew ? "Renews" : "Expires"
            return "\(tier.displayName) • \(renewal) \(formatter.string(from: date))"
        case .expired(_, let date):
            let formatter = DateFormatter()
            formatter.dateStyle = .medium
            return "Expired on \(formatter.string(from: date))"
        case .inGracePeriod(_, let date):
            let formatter = DateFormatter()
            formatter.dateStyle = .medium
            return "Grace period until \(formatter.string(from: date))"
        case .inBillingRetry:
            return "Payment issue - update payment method"
        }
    }
}

@MainActor
class SubscriptionManager: ObservableObject {
    static let shared = SubscriptionManager()

    // MARK: - Published Properties

    @Published private(set) var products: [Product] = []
    @Published private(set) var purchasedProductIDs: Set<String> = []
    @Published private(set) var subscriptionStatus: SubscriptionStatus = .notSubscribed
    @Published private(set) var isLoading = false
    @Published var purchaseError: String?

    // MARK: - Private Properties

    private var updateListenerTask: Task<Void, Error>?

    // MARK: - Initialization

    private init() {
        // Start listening for transactions
        updateListenerTask = listenForTransactions()

        // Load products and check status
        Task {
            await loadProducts()
            await updateSubscriptionStatus()
        }
    }

    deinit {
        updateListenerTask?.cancel()
    }

    // MARK: - Public Methods

    /// Load available products from App Store
    func loadProducts() async {
        isLoading = true

        do {
            let productIDs = SubscriptionTier.allCases.map { $0.rawValue }
            products = try await Product.products(for: productIDs)
                .sorted { $0.price < $1.price }
        } catch {
            print("Failed to load products: \(error)")
        }

        isLoading = false
    }

    /// Purchase a subscription
    func purchase(_ product: Product) async throws {
        isLoading = true
        purchaseError = nil

        do {
            let result = try await product.purchase()

            switch result {
            case .success(let verification):
                let transaction = try checkVerified(verification)
                await transaction.finish()
                await updateSubscriptionStatus()

                // Track analytics
                AnalyticsManager.shared.track(.subscriptionStarted, properties: [
                    "tier": product.id
                ])

            case .userCancelled:
                break

            case .pending:
                purchaseError = "Purchase is pending approval"

            @unknown default:
                break
            }
        } catch {
            purchaseError = "Purchase failed: \(error.localizedDescription)"
            throw error
        }

        isLoading = false
    }

    /// Restore purchases
    func restorePurchases() async throws {
        isLoading = true
        purchaseError = nil

        do {
            try await AppStore.sync()
            await updateSubscriptionStatus()

            AnalyticsManager.shared.track(.subscriptionRestored)
        } catch {
            purchaseError = "Failed to restore purchases: \(error.localizedDescription)"
            throw error
        }

        isLoading = false
    }

    /// Get URL to manage subscription in App Store
    func getManageSubscriptionURL() -> URL? {
        URL(string: "https://apps.apple.com/account/subscriptions")
    }

    /// Open subscription management in App Store
    func openManageSubscription() async {
        guard let scene = UIApplication.shared.connectedScenes.first as? UIWindowScene else { return }

        do {
            try await AppStore.showManageSubscriptions(in: scene)
        } catch {
            // Fall back to URL
            if let url = getManageSubscriptionURL() {
                await UIApplication.shared.open(url)
            }
        }
    }

    /// Check if user has active subscription
    var hasActiveSubscription: Bool {
        subscriptionStatus.isActive
    }

    /// Get formatted price for a product
    func formattedPrice(for product: Product) -> String {
        product.displayPrice
    }

    /// Get renewal disclosure text (required by App Store)
    func renewalDisclosure(for product: Product) -> String {
        let price = product.displayPrice
        let period = product.subscription?.subscriptionPeriod.displayUnit ?? "period"

        return """
        Subscription automatically renews at \(price)/\(period) unless canceled at least 24 hours before the end of the current period. \
        Your account will be charged for renewal within 24 hours prior to the end of the current period. \
        You can manage and cancel your subscription in your App Store account settings after purchase.
        """
    }

    // MARK: - Private Methods

    private func listenForTransactions() -> Task<Void, Error> {
        Task.detached {
            for await result in Transaction.updates {
                do {
                    let transaction = try self.checkVerified(result)
                    await self.updateSubscriptionStatus()
                    await transaction.finish()
                } catch {
                    print("Transaction verification failed: \(error)")
                }
            }
        }
    }

    private func checkVerified<T>(_ result: VerificationResult<T>) throws -> T {
        switch result {
        case .unverified:
            throw StoreError.failedVerification
        case .verified(let safe):
            return safe
        }
    }

    func updateSubscriptionStatus() async {
        var newStatus: SubscriptionStatus = .notSubscribed

        for await result in Transaction.currentEntitlements {
            do {
                let transaction = try checkVerified(result)

                if transaction.productType == .autoRenewable {
                    purchasedProductIDs.insert(transaction.productID)

                    if let tier = SubscriptionTier(rawValue: transaction.productID) {
                        // Check subscription status
                        if let expirationDate = transaction.expirationDate {
                            if transaction.revocationDate != nil {
                                // Subscription was revoked
                                newStatus = .expired(tier: tier, expirationDate: expirationDate)
                            } else if expirationDate > Date() {
                                // Active subscription
                                let status = try? await transaction.subscriptionStatus

                                // Check for grace period or billing retry
                                if let renewalState = status?.first?.state {
                                    switch renewalState {
                                    case .inGracePeriod:
                                        newStatus = .inGracePeriod(tier: tier, expirationDate: expirationDate)
                                    case .inBillingRetryPeriod:
                                        newStatus = .inBillingRetry(tier: tier)
                                    default:
                                        let willRenew = status?.first?.renewalInfo != nil
                                        newStatus = .subscribed(tier: tier, expirationDate: expirationDate, willRenew: willRenew)
                                    }
                                } else {
                                    newStatus = .subscribed(tier: tier, expirationDate: expirationDate, willRenew: true)
                                }
                            } else {
                                newStatus = .expired(tier: tier, expirationDate: expirationDate)
                            }
                        }
                    }
                }
            } catch {
                print("Transaction error: \(error)")
            }
        }

        subscriptionStatus = newStatus
    }
}

// MARK: - Store Errors

enum StoreError: Error, LocalizedError {
    case failedVerification
    case productNotFound
    case purchaseFailed

    var errorDescription: String? {
        switch self {
        case .failedVerification:
            return "Transaction verification failed"
        case .productNotFound:
            return "Product not found"
        case .purchaseFailed:
            return "Purchase failed"
        }
    }
}

// MARK: - Extensions

extension Product.SubscriptionPeriod {
    var displayUnit: String {
        switch unit {
        case .day: return value == 1 ? "day" : "\(value) days"
        case .week: return value == 1 ? "week" : "\(value) weeks"
        case .month: return value == 1 ? "month" : "\(value) months"
        case .year: return value == 1 ? "year" : "\(value) years"
        @unknown default: return "period"
        }
    }
}

// MARK: - Demo Mode Support

extension SubscriptionManager {
    /// Check if we're in demo mode (for App Store review)
    var isDemoMode: Bool {
        PrivacySettings.shared.demoModeEnabled
    }

    /// Check if user should have premium features (subscription OR demo mode)
    var hasPremiumAccess: Bool {
        hasActiveSubscription || isDemoMode
    }
}
